halibot/halauth.py

```diff
@@ -47,20 +57,24 @@ def write_perms(self):
 
 	def grantPermission(self, ri, identity, perm):
 		if not self.enabled:
-			return
+			return False
```

Comment:
We should be raising exceptions instead of the return successful paradigm.

Reply:
Oh, yeah I suppose this would be the case for exceptions here, wouldn't it. Is there an existing exception class I should use here, or should I start inventing my own?

Reply:
Be creative!

Reply:
I'm having second thoughts on this, but for a different reason. Should we allow permissions to be edited when they are not being enforced? Without, there may be a chicken-and-egg problem where you enable permission enforcement, but no one is permitted to modify the permissions table.

Proposal: split `enabled` and `enforced`. Setting `enabled = True` simply enables access to the permissions table internally. Setting `enforced = True` implies `enabled = True`, and is checked in `hasPermission` calls.

Therefore, calls to `grantPermission` when not enabled should be an Exception because you cannot grant a permission when auth is not enabled at all. `True` or `False` still apply for if the permission was added, or already existed.

Reply:
Note, `enabled = True` and `enforced = False` is like selinux "permissive" mode. We should still log failed auth checks. Side note, thinking about adding a "seperate" auth log that logs checks, rejections, etc. May not actually be a separate physical log, but something easy to grep for.